### ISSUE TITLE
Spark SQL Functions Coverage and Parity - Lambda 

### ISSUE THREAD

**Issue by shehabgamin** [lakehq/sail#229]
- [ ] aggregate
- [ ] array_sort
- [ ] exists
- [ ] filter
- [ ] forall
- [ ] map_filter
- [ ] map_zip_with
- [ ] reduce
- [ ] transform
- [ ] transform_keys
- [ ] transform_values
- [ ] zip_with

**Comment by keen85** [comment]
@shehabgamin, are there any plans for tackling this in somewhat near future?

**Comment by shehabgamin** [comment]
> [@shehabgamin](https://github.com/shehabgamin), are there any plans for tackling this in somewhat near future?

@keen85 Would you like to give it a shot?

**Comment by keen85** [comment]
@shehabgamin, I did some analysis which PySpark functions we use in our project and `transform` and some others from this list showed up quite frequently.
So this is blocking us from `sail` adoption.
That is why I ask.

Unfortunately, I can't program Rust myself.

**Comment by shehabgamin** [comment]
@keen85 Got it. Do you happen to know which exact functions are used?

**Comment by keen85** [comment]
@shehabgamin, concerning **lambda** functions?
Several times `pyspark.sql.functions.transform` and a few `pyspark.sql.functions.array_sort`.

**Comment by shehabgamin** [comment]
@keen85 Got it, will take a look and evaluate the scope.

**Comment by shehabgamin** [comment]
@keen85 Looks like we can work with DataFusion to support this upstream hopefully!
https://github.com/apache/datafusion/pull/17220#issuecomment-3577605059

**Comment by shehabgamin** [comment]
Follow-up to ^ here:
https://github.com/apache/datafusion/pull/18921

**Comment by hugopendlebury** [comment]
Datafusion have just merged PR 21679 which adds support for lambdas and includes an initial implementation of transform or as they call it array_transform.

I wouldn't mind having a go at this issue. Would it be possible to give me a high level overview of the starting point and which files / crates I would need to get familiar with ?

**Comment by lonless9** [comment]
Thanks @hugopendlebury . Although the upstream has pushed forward the relevant PR, I guess we won't be able to start the related work until after the datafusion 54.0.0 release in about a month. As for the starting point, I think @davidlghellin  has a better idea.

**Comment by davidlghellin** [comment]
Hi, thanks! 
Yes, I've been working on an initial implementation without using HOFs. I've rebased onto the relevant commit and resolved the conflicts — most of the approach is clear at this point. The part that still needs to be defined is the multi-row handling (how to correctly propagate and evaluate lambda expressions across multiple rows in a vectorized execution context).

For now I'm only testing with filter (PR #1320) as a starting point before generalizing to the other HOFs.

**Comment by twsl** [comment]
@davidlghellin whats the status? do you need help?

**Comment by davidlghellin** [comment]
> [@davidlghellin](https://github.com/davidlghellin) whats the status? do you need help?

@twsl thanks! Quick status:

The higher-order functions are gated on the DataFusion 54 integration that's currently in progress in the repo — DF 54 ships the lambda support the HOFs rely on, so nothing HOF-related can land cleanly until that integration is in.

On the HOF side itself the groundwork is already done on my end: I have a `filter` PR that builds out the lambda machinery (lambda variables + evaluation) and is meant to be the reference pattern the other HOFs (`transform`, `exists`, `aggregate`, `zip_with`, …) plug into. It's essentially ready to go on top of DF 54.

So the sequence is: DF 54 integration lands → filter goes in as the base → the rest fan out from there.

**Comment by twsl** [comment]
is there a timeline or a pr i can follow? couldnt find anything via the search. df 54 apparently has been released two weeks ago

**Comment by davidlghellin** [comment]
> is there a timeline or a pr i can follow? couldnt find anything via the search. df 54 apparently has been released two weeks ago

@twsl yep, here's the PR to follow: https://github.com/lakehq/sail/pull/2057

The DataFusion 54 upgrade has now landed in the repo, so the gate I mentioned is lifted. #2057 builds the `filter` HOF end-to-end (lambda variables + evaluation + distributed serialization) and is the reference pattern the other HOFs plug into.

Sequence from here: `filter` (#2057) goes in as the base → `transform`/`exists`/`forall` fan out next (mechanical on top of this) → `aggregate`/`zip_with`/map-HOFs after.

**Comment by twsl** [comment]
@davidlghellin mind updating the list? i might continue with `map_zip_with` if you arent working on that yet

**Comment by linhr** [comment]
I just updated the list with supported functions.

**Comment by davidlghellin** [comment]
> [@davidlghellin](https://github.com/davidlghellin) mind updating the list? i might continue with `map_zip_with` if you arent working on that yet

@twsl Not started on it, so it's all yours — go ahead with `map_zip_with`.

I'd rather finish closing out the Spark 3.5 gaps I have in flight before picking up anything new here, so it makes more sense for you to take it than to have it sit on my plate.
